queries/period_slippage.sql

```diff
@@ -310,70 +310,139 @@ incoming_and_outgoing_with_buffer_trades as (
+        sum(amount) token_imbalance_wei,
+        symbol,
+        token,
+        tx_hash,
+        date_trunc('hour', block_time) as hour
+    from
+        incoming_and_outgoing_with_buffer_trades
+    group by
+        symbol,
+        token,
+        solver_address,
+        solver_name,
+        tx_hash,
+        block_time
+    having
+        sum(amount) != 0
 ),
-
-exact_prices as (
+token_times as (
+    select
+        hour,
+        token
+    from
+        final_token_balance_sheet
+    group by
+        hour,
+        token
+),
+precise_prices as (
     select
-        token as contract_address,
+        contract_address,
         decimals,
-        price
-    from final_token_balance_sheet
-    left join prices.usd
-        on contract_address = token
-        and minute = '{{EndTime}}'
+        date_trunc('hour', minute) as hour,
+        avg(price) as price
+    from
+        prices.usd pusd
+    inner join token_times tt
+        on minute between date(hour) and date(hour) + interval '1 day' 
+        and date_trunc('hour', minute) = hour
+        and contract_address = token
+    group by
+        contract_address,
+        decimals,
+        date_trunc('hour', minute) 
 ),
--- The approximate prices are based on the hourly median prices over the past day.
-approximate_prices as (
+median_prices as (
     select
         contract_address,
         decimals,
-        percentile_cont(0.5) WITHIN group (order by median_price) as price
-    from prices.prices_from_dex_data p
-    where hour > '{{EndTime}}'::timestamptz - interval '1 day'
-    and date(hour) = '{{EndTime}}'::timestamptz - interval '1 day'
-      -- Getting approximate prices only for tokens we do not have an exact price for.
-    and contract_address IN (select contract_address from exact_prices where price is null)
-    group by contract_address, decimals
-    having sum(sample_size) > 0
+        tt.hour,
+        median_price
+    from
+        prices.prices_from_dex_data musd
+        inner join token_times tt on musd.hour = tt.hour
+        and contract_address = token
 ),
-end_prices as (
-    select * from exact_prices where price is not null
-    union
-    select * from approximate_prices
+intrinsic_prices as (
+    select 
+        contract_address,
+        decimals,
+        hour,
+        AVG(price) as price
+    from (
+        select
+            buy_token_address as contract_address,
+            ROUND(LOG(10, atoms_bought / units_bought)) as decimals,
+            date_trunc('hour', block_time) as hour,
+            trade_value_usd / units_bought as price
+        FROM gnosis_protocol_v2."trades"
+        WHERE units_bought > 0
+    UNION
+        select
+            sell_token_address as contract_address,
+            ROUND(LOG(10, atoms_sold / units_sold)) as decimals,
+            date_trunc('hour', block_time) as hour,
+            trade_value_usd / units_sold as price
+        FROM gnosis_protocol_v2."trades"
+        WHERE units_sold > 0
```

Comment:
could we add:
WHERE units_sold > 0 
         and b.block_time between '{{StartTime}}' and '{{EndTime}}'
        case
              when '{{TxHash}}' = '0x' then true
              else replace('{{TxHash}}', '0x', '\x') :: bytea = tx_hash
        end
        
        in order to speed it up in case we only look for a tx.

Reply:
While this looks good - This constraint would need to be in both sections of the union here. We should do this in a separate PR.